File: src/main/java/ch/obermuhlner/astro/image/WriteThroughArrayDoubleImage.java
package ch.obermuhlner.astro.image;

import ch.obermuhlner.astro.image.color.ColorModel;

public class WriteThroughArrayDoubleImage implements DoubleImage {

  private final DoubleImage image;
  private final ArrayDoubleImage arrayImage;

  public WriteThroughArrayDoubleImage(DoubleImage image, ColorModel colorModel) {
    this.image = image;

    this.arrayImage = new ArrayDoubleImage(image.getWidth(), image.getHeight(), colorModel);
  }

  @Override
  public int getWidth() {
    return arrayImage.getWidth();
  }

  @Override
  public int getHeight() {
    return arrayImage.getHeight();
  }

  @Override
  public double[] getNativePixel(int x, int y, double[] color) {
    return arrayImage.getNativePixel(x, y, color);
  }

  @Override
  public void setNativePixel(int x, int y, double[] color) {
    arrayImage.setNativePixel(x, y, color);
    image.setNativePixel(x, y, color);
  }
}

File: src/main/java/ch/obermuhlner/astro/image/WriteThroughArrayDoubleImage.kt
package ch.obermuhlner.astro.image

import ch.obermuhlner.astro.image.color.ColorModel

class WriteThroughArrayDoubleImage(private val image: DoubleImage, colorModel: ColorModel) : DoubleImage {
    private val arrayImage: ArrayDoubleImage = ArrayDoubleImage(image.width, image.height, colorModel)

    override val width: Int
        get() = arrayImage.width
    override val height: Int
        get() = arrayImage.height

    override fun getNativePixel(x: Int, y: Int, color: DoubleArray?): DoubleArray {
        return arrayImage.getNativePixel(x, y, color)
    }

    override fun setNativePixel(x: Int, y: Int, color: DoubleArray) {
        arrayImage.setNativePixel(x, y, color)
        image.setNativePixel(x, y, color)
    }

}